### PR TITLE
8337199: Add jcmd Thread.vthread_scheduler and Thread.vthread_pollers diagnostic commands

### PR DESCRIPTION
**Pull request by AlanBateman**
Adds `jcmd <pid> Thread.vthread_scheduler` to print the virtual thread scheduler and `jcmd <pid> Thread.vthread_pollers` to print the I/O pollers that support virtual threads doing blocking network I/O operations.

This is a subset of the diagnostics that we've had in the loom repo for a long time. @larry-cable proposed a PR recently ([pull/22121](https://github.com/openjdk/jdk/pull/22121)) to bring a version of same into main line but it was based on an older proposal. This new PR supplants that effort.

The jtreg failure handler is updated to execute Thread.vthread_scheduler, useful capture when a test fails or times out.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8339420](https://bugs.openjdk.org/browse/JDK-8339420) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8337199](https://bugs.openjdk.org/browse/JDK-8337199): Add jcmd Thread.vthread_scheduler and Thread.vthread_pollers diagnostic commands (**Enhancement** - P4)
 * [JDK-8339420](https://bugs.openjdk.org/browse/JDK-8339420): Add jcmd Thread.vthread_scheduler and Thread.vthread_pollers diagnostic commands (**CSR**)

### Reviewers
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - **Reviewer**) Review applies to [2e9619a1](https://git.openjdk.org/jdk/pull/22414/files/2e9619a1fe2f11bfd4a3adcf7b572ca80c1c6c56)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22414/head:pull/22414` \
`$ git checkout pull/22414`

Update a local copy of the PR: \
`$ git checkout pull/22414` \
`$ git pull https://git.openjdk.org/jdk.git pull/22414/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22414`

View PR using the GUI difftool: \
`$ git pr show -t 22414`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22414.diff">https://git.openjdk.org/jdk/pull/22414.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22414#issuecomment-2505769676)
</details>

### CODE REVIEW

File: test/hotspot/jtreg/serviceability/dcmd/thread/VThreadCommandsTest.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2023, 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (liach) [inline test/hotspot/jtreg/serviceability/dcmd/thread/VThreadCommandsTest.java:2]:
```suggestion
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
```

File: test/hotspot/jtreg/serviceability/dcmd/thread/VThreadCommandsTest.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+
+class VThreadCommandsTest {
+
+    /**
+     * Thread.vthread_scheduler
+     */
+    @Test
+    void testVThreadScheduler() {
+        // ensure default scheduler are timeout schedulers are initialized
```
Comment (liach) [inline test/hotspot/jtreg/serviceability/dcmd/thread/VThreadCommandsTest.java:62]:
```suggestion
        // ensure default scheduler and timeout schedulers are initialized
```

File: test/hotspot/jtreg/serviceability/dcmd/thread/VThreadCommandsTest.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                    }
+                }
+                return null;
+            }).get();
+        }
+
+        jcmd("Thread.vthread_pollers")
+                .shouldContain("Read I/O pollers:")
+                .shouldContain("Write I/O pollers:")
+                .shouldContain("[0] sun.nio.ch");
```
Comment (liach) [inline test/hotspot/jtreg/serviceability/dcmd/thread/VThreadCommandsTest.java:96]:
```suggestion
                .shouldContain("[0] sun.nio.ch.");
```
As unlikely as this is, I recommend always use a trailing separator for package names - there are corner cases like `java.lang.ref` vs `java.lang.reflect`

File: test/hotspot/jtreg/serviceability/dcmd/thread/VThreadCommandsTest.java (hunk 1 above)
Comment (AlanBateman) [reply test/hotspot/jtreg/serviceability/dcmd/thread/VThreadCommandsTest.java:2]: The test initially created in Oct 2023. It might look like a new test in main line but I think we are supposed to keep the original creation date.

File: test/hotspot/jtreg/serviceability/dcmd/thread/VThreadCommandsTest.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                    }
+                }
+                return null;
+            }).get();
+        }
+
+        jcmd("Thread.vthread_pollers")
+                .shouldContain("Read I/O pollers:")
+                .shouldContain("Write I/O pollers:")
+                .shouldMatch("^\\[0\\] sun.nio.ch..+ \\[registered = [\\d]+, owner = .+\\]$");
```
[the remark this answers is not in the document]
Comment (AlanBateman) [reply test/hotspot/jtreg/serviceability/dcmd/thread/VThreadCommandsTest.java]: Well spotted, these should be escaped although it not doing so won't impact the match here.

[the remark this answers is not in the document]
Comment (kevinjwalls) [reply test/hotspot/jtreg/serviceability/dcmd/thread/VThreadCommandsTest.java]: Yes -it was almost the definitive bikeshedding comment.

**Review by kevinjwalls — APPROVED** [review]

**Review by dholmes-ora — COMMENTED** [review]
The two new `execute` methods differ only in the Symbol naming the method to call. Can we not at least define a static helper in the file so that we don't need to duplicate everything?

**Comment by AlanBateman** [comment]
> The two new `execute` methods differ only in the Symbol naming the method to call. Can we not at least define a static helper in the file so that we don't need to duplicate everything?

Okay, esp. as we'll likely add more. At some point I think Larry is doing to replace all this with other infrastructure for commands in Java.

**Review by dholmes-ora — APPROVED** [review]
LGTM! Thanks
